File: Leetcode/BFS/#133. Clone Graph.cpp
//BFS
class Solution {
public:
    Node* cloneGraph(Node* node) {
        if(node==NULL) return NULL;
        unordered_map<Node*, Node*>mp;
       
        mp[node]=new Node(node->val);
        queue<Node*>Q;
        Q.push(node);
        
        while(!Q.empty()){
            auto cur = Q.front(); Q.pop();
            for(auto neighbor: cur->neighbors){
                if(!mp.count(neighbor)){
                mp[neighbor]=new Node(neighbor->val);
                Q.push(neighbor);
                }
                
                mp[cur]->neighbors.push_back(mp[neighbor]);
            }
        }
        
        return mp[node];
    }
};

//DFS -> key point: if the current node is visited, don't perform dfs but just return previously determined Node*

class Solution {
public:
    
    Node* dfs(Node* node,unordered_map<Node*,Node*>&mp){
       if(!node) return nullptr;
        
        if(!mp.count(node)){ // <- this is the key point
            mp[node]=new Node(node->val);
         for(auto &neighbor: node->neighbors){
            mp[node]->neighbors.push_back(dfs(neighbor,mp));
            }
        }
        
        return mp[node];
    }
    Node* cloneGraph(Node* node) {
        unordered_map<Node*, Node*>mp;
        return dfs(node,mp);
    }
};

File: Graph/#133. Clone Graph.py
''' Method 1 - DFS
TC: O(V + E)
SC: O(V)
'''
class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        if not node:
            return None
        self.copy = dict()
        return self.dfs(node)


    # Returns copy of original node
    def dfs(self, node):
        if node.val in self.copy:
            return self.copy[node.val]

        self.copy[node.val] = Node(node.val)
        for neighbor in node.neighbors:
            self.copy[node.val].neighbors.append(self.dfs(neighbor))
        
        return self.copy[node.val]
    
''' Method 2 - BFS
TC: O(V + E)
SC: O(V)
'''

class Solution:
    def cloneGraph(self, node: 'Node') -> 'Node':
        if not node:
            return node
            
        Q = deque([node])
        copy = {node.val: Node(node.val)}
        
        while Q:
            curr = Q.popleft()
            for neighbor in curr.neighbors:
                if neighbor.val not in copy:
                    copy[neighbor.val] = Node(neighbor.val)
                    Q.append(neighbor)
                copy[curr.val].neighbors.append(copy[neighbor.val])

        return copy[node.val]
